### ISSUE TITLE
Parsing dataclass from file doesn't work with required=True or default value

### ISSUE THREAD

**Issue by RoryDungan** [omni-us/jsonargparse#54]
Hi! I am attempting to parse a config file into an namespace, such that the whole contents of the config YAML file is assigned to that namespace. I'm using a dataclass that matches the format of the YAML file.

I've found two ways to do this:
```
parser.add_argument('--server-config', type=ServerConfig)
# or
parser.add_dataclass_arguments(ServerConfig', 'server-config')
```
However, there doesn't seem to be a way to provide a default value or mark the argument as required. If I try to provide a default via `default='server-config.yaml'` I get a ValueError saying the string is not an instance of the class:
```
ValueError: Expected "default" argument to be an instance of "ServerConfig" or its kwargs dict, given server-config.yaml
```
And if I try to mark it as required then I always get an error saying the field is required but not included, even when I pass `--server-config server-config.yaml` on the command-line. 

Is using the library in this way supported and if so, how might I provide a default value or make it required? Is the only way to provide a default config file and mark options as required to define everything explicitly, as opposed to using dataclasses?

**Comment by mauvilsa** [comment]
The dataclass support is relatively new so there might be some edge cases not considered yet. Regarding the default value, you could do:

```python
from dataclasses import dataclass
from jsonargparse import ArgumentParser

@dataclass
class ServerConfig:
    param1: int = 1
    param2: float = 2.3
    param3: str = '4'

parser = ArgumentParser()
parser.add_argument('--server-config', type=ServerConfig, default=ServerConfig())
cfg = parser.parse_args(with_meta=False)
print(cfg)
```

which if executed would print `Namespace(server_config=Namespace(param1=1, param2=2.3, param3='4'))`. Then if you have a yaml file with contents:

```yaml
param1: 5
param2: 6.7
param3: '8'
```

and then run it as `tool.py --server-config server-config.yaml` it would print `Namespace(server_config=Namespace(param1=5, param2=6.7, param3='8'))`.

If instead the argument is added as `parser.add_argument('--server-config', type=ServerConfig, required=True)` and run the tool without any arguments it gives `Namespace(server_config=Namespace(param1=None, param2=None, param3=None))` which is weird since it is required and should fail. I will need to check what can be done here.

Regarding the `add_dataclass_arguments` method there are unit tests at https://github.com/omni-us/jsonargparse/blob/master/jsonargparse_tests/signatures_tests_gt_py35.py#L43-L79. I see there is nothing related to required so I will also need to check that.

**Comment by RoryDungan** [comment]
Thanks for that! The default values in dataclasses are very helpful as are `default_config_files`, although looking at the docs I can't see a way to have a namespaced default config file. From what I can tell, using a dataclass is the only way to have a whole config file go into a namespace. 

This isn't a blocking issue for me and I'm thinking of just working around it by including the namespace in the config file like so:
```yaml
server_config:
  param1: 5
  param2: 6.7
  param3: '8'
```
This means I can't pass the same config file directly to the `server-config` namespace via command-line any more, but that's not a big deal since I can just add another argument with `ActionConfigFile` to allow users to specify the config file at the root level.

**Comment by RoryDungan** [comment]
On second thought I think there are two separate issues I'm having here:
 - When using dataclasses, there's no way to mark fields as required. I'm not sure how you'd do this per-field but it would make sense to be able to mark the whole dataclass as required and have that make all its fields required.
 - The only way to load a config file into a namespace appears to be using a dataclass, but there's no way to supply a default config file to load since `default_config_files` only affects root-level config files. This isn't a big issue for me as there's the easy workaround I mentioned before, and I'm wondering now if maybe the way I was trying to use this goes against the indended design of the library.

**Comment by mauvilsa** [comment]
> I can't see a way to have a namespaced default config file

Indeed default config files are global.

> This means I can't pass the same config file directly to the server-config namespace via command-line any more

Yes, you can. This is where the precedence comes in. Default config files are for defaults, imagine files in `/etc` or in `~/.myapp/`. But providing arguments or config files via command line will override these. So `tool.py --server-config server-config.yaml` works as you want.

> it would make sense to be able to mark the whole dataclass as required and have that make all its fields required.

This is one of the options I was thinking of. But still need to look into this. Anyway note that if an argument has a default, even if it is marked as required it will not fail because the default value is there.

> The only way to load a config file into a namespace appears to be using a dataclass

Apart from the signature methods (class, method, function, dataclass) there is also `ActionJsonSchema`.

> I'm wondering now if maybe the way I was trying to use this goes against the indended design of the library

There are many ways to use the library. Until now I don't see any weird usage. Only learning how to use it. Certainly there could be bugs that should be fixed.

**Comment by mauvilsa** [comment]
To make individual arguments required, the parameters should not have a default. This is how it was designed and it worked correctly in methods implemented initially: `add_class_arguments`, `add_method_arguments` and `add_function_arguments`. But there was a mistake in the new `add_dataclass_arguments` which meant that this was not being done correctly. This is fixed in v3.11.1 just released. Using the example above, for instance to make `param1` required it would be simply:

```
@dataclass
class ServerConfig:
    param1: int
    param2: float = 2.3
    param3: str = '4'
```

The `required` option that `add_dataclass_arguments` had was also a mistake. The only signature method that receives a required is `add_subclass_arguments`. This is also fixed in v3.11.1.

**Comment by mauvilsa** [comment]
Closing this. @RoryDungan please reopen if after the fix and explanation it does not work as expected.
